### DISCUSSION TITLE
What about multiple configuration files?

### DISCUSSION THREAD

**Discussion by ostap667inbox** [Mikubill/sd-webui-controlnet#705] [Ideas]
**Introduction:**

- Right now we have _7 different_ configuration files for Control Net and t2i adapters and more then 30 different model files. We are also expecting soon to have a request to add a model working with faces - this will add two more model files and two more configuration files (versions for SD1.5 and SD2.1).
- Also, having such a convenient and powerful tool like Multi-ControlNet, where you can use different combinations of models, we came in the end to the fact that the existing two text fields in the 'Settings', where you enter manually one configuration file for Control Net and one for Adapter, was not enough and in addition to constantly getting into settings very inconvenient.
_Example: what do I do if I want to use t2iadapter_keypose and t2iadapter_style on two adjacent Multi-ControlNet tabs - these models require different configuration files._

**What would be better to change?**

1. The most obvious solution would be to add an updatable drop-down menu to each Multi-ControlNet tab, where you can select a configuration file manually, and change the two text boxes in the settings to one: the path to the directory where the configuration files are located.
This is also a more universal way, because each user may have a different set of downloaded model files and configurations.

2. However, an even more convenient solution would be to have an automatic mapping of the configuration file for each selected model, that is, in addition to the models and configuration files themselves, have another file listing the mappings.
This method is more attractive because it completely removes one control from the complex interface, but it requires the user to understand that when loading a model file, he must also manually edit the file in which the mappings of models to their configuration are prescribed.

**Shall we implement one of the suggested ways? Discuss?**

**Comment by ljleb** [comment]
If detecting the configuration is possible, we should do this in the long run. I do not know about this so we need the feedback of someone more knowledgable.

In the meantime we could look into implementing a dropdown for config files, I think it is a reasonable solution in the absence of knowledge on how to detect configuration automatically.

By default we could also (or only) look for config files that have the same name as the selected model. This would allow to automate the process in some way.

**Reply by ostap667inbox** [reply]
> By default we could also (or only) look for config files that have the same name as the selected model. This would allow to automate the process in some way.

This is also a very good solution. True, the number of configuration files will greatly increase and users will have to worry about renaming existing files. However, this is a small price to pay if the result is full automatization.

**Comment by ostap667inbox** [comment]
Hmm. Is that possible right now?

_[lllyasviel](https://github.com/lllyasviel) commented [5 hours ago](https://github.com/Mikubill/sd-webui-controlnet/pull/688#issuecomment-1494182277)
this webui plugin can read overwritten configs. Please consider copy cldm_v21.yaml to "controlnet_sd21_laion_face_v2_prund.yaml" and then put it in the same folder with the model so that users do not need to set settings manually._

If so, that's great, we should add a mention of this in the readme.md

**Reply by ostap667inbox** [reply]
https://github.com/Mikubill/sd-webui-controlnet/pull/688#issuecomment-1496411566

> > I have two questions.
> > 
> > 1. Did I get it right that it is enough for the ControlNet plugin if the folder with the models contains the same number of yaml-files as the models and has the same names as the model files? And that is why it is possible to use models with different configuration files in Multi-ControlNet at the same time, e.g. t2iadapter_keypose and t2iadapter_style?
> > 2. Is it necessary to leave these two text fields blank in 'Settings' in this case? If these two text boxes are filled in, but at the same time the configuration files are renamed according to the model names, which will have the highest priority?
> 
> Yes that's correct. As long as each model has a corresponding yaml with the same name, it will use that rather than the default in the settings. You can leave the defaults in the settings alone.
